docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,447 @@
+- Introduce capabilities to reconcile add-ons in an order that respects dependencies between add-ons.
+- Deprecate or remove the ClusterResourceSet experimental feature.
+
+## Proposal
+
+This document introduces the concept of a ClusterAddonProvider, a new pluggable component to be deployed on the Management Cluster. A ClusterAddonProvider must be implemented using a specified package management tool and will act as an intermediary/proxy between Cluster API and the chosen package management solution.
+
+This proposal also introduces a ClusterAddonProvider for [Helm](https://helm.sh/) as a concrete example of how a ClusterAddonProvider would work with a package management tool. Users can configure ClusterAddonProvider for Helm through a new Kubernetes CRD called HelmChartProxy to be deployed on the Management Cluster. HelmChartProxy will act as an intermediary (proxy) between Cluster API and the Helm to install packages onto workload clusters. The implementation of ClusterAddonProvider for Helm can be used as a reference to implement ClusterAddonProviders for other package managers, i.e. ClusterAddonProvider for kapp.
+
+### User Stories
+
+#### Story 1
+
+As a developer or Cluster operator, I would like Cluster add-on providers to automatically include or install the package management tool.
+
+#### Story 2
+
+As a developer or Cluster operator, I would like Cluster add-on providers to expose configurations native to the package management tool to install Cluster add-ons into the newly created Workload Cluster.
+
+#### Story 3
+
+As a developer or Cluster operator, I would like Cluster add-on providers to leverage existing capabilities of the package management tool, such as identifying the proper version of Cluster add-ons package to install onto the workload Cluster.
+
+#### Story 4
+
+As a developer or Cluster operator, I would like Cluster add-on providers to use information available in the Management Cluster (e.g. service or pod CIDR on the Cluster object) for customizing add-ons configurations specifically for the newly created workload Cluster.
+
+#### Story 5
+
+As a developer or Cluster operator, I would like Cluster add-on providers to update add-ons configurations when information available in the Management Cluster (e.g add-on configuration) changes.
+
+### ClusterAddonProvider Functionality
+
+A ClusterAddonProvider is a new component for bringing add-ons into the Cluster API lifecycle. It serves as a general solution to Cluster API Add-ons with multiple concrete implementations, similar to the relationship between Cluster API and infrastructure providers. It aims to do so handle add-on lifecycle management in a holistic way by providing the following capabilities:
+
+- Provide or install a package management tool
+- Determine which add-ons to install or uninstall on each Cluster
+- Provide a default configuration for each add-on
```

Comment:
I think that was part of @fabriziopandini's original requirements but I believe the intention was to ensure that if we supply a blank `valuesTemplate`, we will install a Helm chart using the default values.yaml instead of setting them all to empty fields.

Reply:
Makes sense. As far as I'm aware with Helm that's just how it works per default. Even if you provide an empty values.yaml it would just use the default from the Helm Chart.

Reply:
imagine a user that wants to customize default values in a helm chart, this goal is about making this simple by allowing to do the customization one instead of for each cluster.

Reply:
Ah now I've got it :). So essentially the valuesTemplate in the HelmChartProxy in the current implementation?

Reply:
Yep pretty much.

Reply:
Alright. Fine to resolve for me